File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        def rec(l1, l2):
            if l1<=0 or l2<=0: return 0
            if dp[l1][l2]!=-1: return dp[l1][l2]
            
            if text1[l1-1]==text2[l2-1]: dp[l1][l2] = rec(l1-1, l2-1)+1
            else: dp[l1][l2] = max(rec(l1, l2-1), rec(l1-1, l2))
            return dp[l1][l2]
        
        dp = [[-1]*(len(text2)+1) for _ in range(len(text1)+1)]
        return rec(len(text1), len(text2))



# class Solution:
#     def longestCommonSubsequence(self, text1: str, text2: str) -> int:
#         def rec(l1, l2):
#             if l1<=0 or l2<=0: return 0
#             if text1[l1-1]==text2[l2-1]:
#                 return rec(l1-1, l2-1)+1
#             return max(rec(l1, l2-1), rec(l1-1, l2))
        
#         return rec(len(text1), len(text2))

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int ln1=text1.length(), ln2=text2.length();
        int[] prev = new int[ln2+1];
        
        for(int l1=1; l1<=ln1; ++l1){
            int[] dp = new int[ln2+1];
            for(int l2=1; l2<=ln2; ++l2){
                dp[l2] = text1.charAt(l1-1) == text2.charAt(l2-1) ? prev[l2-1]+1 : Math.max(prev[l2], dp[l2-1]);
            }
            prev = dp;
        }
        
        return prev[ln2];
    }
}



// class Solution {
//     public int longestCommonSubsequence(String text1, String text2) {
//         int ln1=text1.length(), ln2=text2.length();
//         int[][] dp = new int[ln1+1][ln2+1];
        
//         for(int l1=1; l1<=ln1; ++l1){
//             for(int l2=1; l2<=ln2; ++l2){
//                 dp[l1][l2] = text1.charAt(l1-1)==text2.charAt(l2-1) ? 1+dp[l1-1][l2-1] : Math.max(dp[l1-1][l2], dp[l1][l2-1]);
//             }
//         }
        
//         return dp[ln1][ln2];
//     }
// }